.tekton/pull-request.yaml

```diff
@@ -46,6 +46,22 @@ spec:
             - name: run-tests
               image: registry.access.redhat.com/ubi8/ubi:latest
               script: cd $(workspaces.workspace.path) && ./automation/test.sh
+      - name: run-gitlint
+        workspaces:
+          - name: workspace
+            workspace: workspace
+        runAfter:
+          - clone-repository
+        taskSpec:
+          workspaces:
+            - name: workspace
+          steps:
+            - name: run-gitlint
+              image: registry.access.redhat.com/ubi9/python-39:latest
+              script: |
+                python -m pip install gitlint
+                cd $(workspaces.workspace.path)
+                gitlint --commits $(git merge-base HEAD origin/main)..HEAD
```

Comment:
@amisstea I'm trying here with tekton, as an alternative to github actions. The problem is that it should fail but it's passing and I'm a bit limited with debugging here.

The main issue is that I don't have the logs on one hand, and on the other hand, I don't know which branches are being pulled using the previous task, so it's hard to replicate locally. 

Any suggestions?

Reply:
I tested this out on my fork and discovered an error from the tekton admission controller because the `params` aren't defined in this embedded task.

```shell
{"severity":"ERROR","timestamp":"2023-03-03T17:09:28.691089419Z","logger":"pipelinesascode","caller":"events/emit.go:45","message":"PipelineRun on-pull-request- has failed: creating pipelinerun on-pull-request- in amisstea has failed: admission webhook \"validation.webhook.pipeline.tekton.dev\" denied the request: validation failed: non-existent variable in \"python -m pip install gitlint\\ncd $(workspaces.workspace.path)\\ngitlint --commits $(git merge-base HEAD \\\"$(params.git-url)/$(params.target-branch)\\\")..HEAD\\n\": spec.pipelineSpec.tasks[2].taskSpec.steps[0].script ","commit":"fdfd333","provider":"github","event-id":"273ae060-b9e6-11ed-8b5f-fc68c01af0fc","event-sha":"a5c0d170e192396b804c126550e373b941db787a","event-type":"pull_request","stacktrace":"github.com/openshift-pipelines/pipelines-as-code/pkg/events.(*EventEmitter).EmitMessage\n\t/src/pkg/events/emit.go:45\ngithub.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode.(*PacRun).Run.func1\n\t/src/pkg/pipelineascode/pipelineascode.g...
```

They need to be declared in both the inline `Task` as well as the underlying `taskRef`. Here's an [example](https://github.com/amisstea/o11y/blob/218c26b451f17c4aa7c935120cc3aa38fe69e6fc/.tekton/pull-request.yaml#L64).

To debug this I used the `tkn-pac` CLI to setup a personal GitHub app which triggers PaC on my cluster. I also had to create a `Repository` CR in the desired namespace. This can all be accomplished using `tkn-pac`.

```shell
# Execute these commands from your fork
tkn-pac bootstrap # Follow the prompts
tkn-pac create repository # Decline the webhook setup during this step
```
A couple things to note:
1. I'm using a hypershift cluster so there was no need to [setup a proxy](https://pipelinesascode.com/docs/install/installation/#proxy-service-for-pac-controller) to get the PaC webhook working.
2. My example above doesn't perform `gitlint` on all the new commits in a pull request, just the head commit. That needs more consideration.
3. Be careful with the embedded subshell you have here (`git merge-base`) because errors from it won't bubble up to the main process. You can work around this by evaluating it on a separate line.

Reply:
@gbenhaim do you know of anything we can do to make debugging CI issues like this easier with the shared infra? It seems like PaC doesn't produce meaningful errors (or anything for that matter) when there are syntax errors in a `PipelineRun`. Also, I didn't have access to the cluster so I'm assuming the rest of the team will run into this limitation as well.

Reply:
@amisstea 
> ```shell
> # Execute these commands from your fork
> tkn-pac bootstrap # Follow the prompts
> tkn-pac create repository # Decline the webhook setup during this step
> ```

Is that my olly fork? infra-deployments? I tried with both and am getting:
```
→ tkn-pac bootstrap
🏃 Checking if Pipelines as Code is installed.
Error: Unauthorized
```

Reply:
Currently the pipeline fails, while I'm expecting it to pass.
This is what I'm doing at the moment:
```
              script: |
                #!/bin/bash -ex

                python -m pip install gitlint
                git remote add upstream "$(params.git-url)"
                git fetch upstream "$(params.target-branch)"
                BASELINE_COMMIT="$(git merge-base HEAD upstream/"$(params.target-branch)")"
                gitlint --commits "$BASELINE_COMMIT"..HEAD
```
In the little logs I'm getting, I'm seeing this:
```
* [new branch]      main       -> upstream/main
++ git merge-base HEAD upstream/main
+ BASELINE_COMMIT=
```
It seems to me as merge-base fails but I cannot tell why.

Reply:
Also, I think merge-base is going to be a problem here in case the PR branch is not rebased on the target branch when this is running, but maybe it will just point to an older commit and won't be a problem.

Reply:
I tested a copy of your branch with my setup and managed to get it [working](https://github.com/amisstea/o11y/pull/4/files). The main issue I found is that the `git-clone` task is performing a shallow clone with `depth`=1 by default. I don't see a way to disable that from the task params (I tried `depth`=0) but there's a simple workaround. I've tested this with a couple malformed commit messages and it raised errors for both with no need for `git merge-base`. Here's the gist of it:

```bash
git fetch --unshallow
gitlint --commits "origin/$(params.target-branch)..HEAD"
```